File: ds practical/cpp/arrays and searching sorting algo/library_managment.cpp
#include <iostream>
#include <string>
using namespace std;

// Function to compute average borrowed books
float averageBorrowed(int borrowed[], int n) {
    int total = 0;
    for (int i = 0; i < n; i++)
        total += borrowed[i];
    return (float)total / n;
}

// Function to find highest and lowest borrowed books
void findHighLow(string books[], int count[], int m) {
    int maxIndex = 0, minIndex = 0;
    for (int i = 1; i < m; i++) {
        if (count[i] > count[maxIndex]) maxIndex = i;
        if (count[i] < count[minIndex]) minIndex = i;
    }
    cout << "Highest borrowed book: " << books[maxIndex] << endl;
    cout << "Lowest borrowed book: " << books[minIndex] << endl;
}

// Function to count members with 0 borrowings
int countZero(int borrowed[], int n) {
    int c = 0;
    for (int i = 0; i < n; i++)
        if (borrowed[i] == 0) c++;
    return c;
}

// Function to find most frequent borrow count (mode)
void mostFrequent(string books[], int count[], int m) {
    int freq[100] = {0};  // frequency array
    for (int i = 0; i < m; i++)
        freq[count[i]]++;

    int modeVal = count[0];
    for (int i = 1; i < m; i++) {
        if (freq[count[i]] > freq[modeVal])
            modeVal = count[i];
    }

    cout << "Most frequent borrow count: " << modeVal << endl;
    cout << "Books with this count:" << endl;
    for (int i = 0; i < m; i++) {
        if (count[i] == modeVal)
            cout << books[i] << endl;
    }
}

int main() {
    string members[] = {"Alice", "Bob", "Charlie", "David", "Eve", "Frank"};
    int borrowed[] = {5, 0, 3, 5, 2, 0};
    int n = 6;

    string books[] = {"Book A", "Book B", "Book C", "Book D", "Book E"};
    int count[] = {10, 4, 7, 1, 4};
    int m = 5;

    cout << "Average books borrowed: " << averageBorrowed(borrowed, n) << endl;
    findHighLow(books, count, m);
    cout << "Members with 0 books: " << countZero(borrowed, n) << endl;
    mostFrequent(books, count, m);

    return 0;
}

File: ds practical/py/arrays and searching sorting/library_managment.py
# Library Borrowing Management System

# Sample data: dictionary of members and the number of books they borrowed
members_borrow = {
    "Alice": 5,
    "Bob": 0,
    "Charlie": 3,
    "David": 5,
    "Eve": 2,
    "Frank": 0
}

# Sample data: dictionary of books and how many times they have been borrowed
book_borrow_count = {
    "Book A": 10,
    "Book B": 4,
    "Book C": 7,
    "Book D": 1,
    "Book E": 4
}

# 1️⃣ Compute average number of books borrowed by all members
def average_books_borrowed(borrow_data):
    total = sum(borrow_data.values())
    count = len(borrow_data)
    avg = total / count
    return avg

# 2️⃣ Find book with highest and lowest number of borrowings
def find_high_low_borrowed_books(book_data):
    highest = max(book_data, key=book_data.get)
    lowest = min(book_data, key=book_data.get)
    return highest, lowest

# 3️⃣ Count members who have not borrowed any books
def count_non_borrowers(borrow_data):
    count = sum(1 for x in borrow_data.values() if x == 0)
    return count

# 4️⃣ Find the most frequently borrowed book count (mode of borrow counts)
from collections import Counter
def most_frequent_borrow_count(book_data):
    freq = Counter(book_data.values())
    mode_count = freq.most_common(1)[0][0]
    # Return all books having that borrow count
    most_frequent_books = [book for book, cnt in book_data.items() if cnt == mode_count]
    return most_frequent_books, mode_count

# ---- MAIN EXECUTION ----
print("Average books borrowed per member:", average_books_borrowed(members_borrow))

high, low = find_high_low_borrowed_books(book_borrow_count)
print(f"Most borrowed book: {high}")
print(f"Least borrowed book: {low}")

print("Number of members who borrowed no books:", count_non_borrowers(members_borrow))

books, count = most_frequent_borrow_count(book_borrow_count)
print(f"Most frequent borrow count: {count}, Books: {books}")
